### PR TITLE
islice_extended function

### PR DESCRIPTION
**Pull request by bbayles**
Re: [this discussion](https://github.com/erikrose/more-itertools/pull/145#issuecomment-305239617), this PR adds `islice_extended()`, which is like `islice()`, but without the restrictions on the arguments.

The goal here is to do slices with negative arguments without needing to store the entire iterable. This is generally possible, but a bit tricky since we don't know the length of the iterable.

Some cases are easier than others. I wrote them all out and then consolidated ones with similar logic. Below I'll sketch out one, but I think the comments in the code should describe the others pretty well.

---

To work through one example (positive start, negative stop, positive step), suppose we have `islice_extended(iter('ABCDEFGH'), 2, -2, 1)` we'll want to wind up with this (blue means yielded):
image

We can start by advancing to the `start` position; we don't need anything before that:
image

We don't know where `-2` is, so we'll cache two items in `deque` (orange means cached):
image

As we iterate through we'll pop the leftmost item in our cache and add the current item to the right side of the cache:
image
image
image
image

---

As the CPython docs [say](https://github.com/python/cpython/blob/2ddf5a19c3a06978edff2c8ba0aaf5df3528204a/Objects/sliceobject.c#L198), slice indexing is hard to get right! I *think* I've covered the bases here, but watchers, tell me if you see something amiss.

### CODE REVIEW

**Comment by erikrose** [comment]
This is a clever and very itertoolsy approach. :-)

**Comment by bbayles** [comment]
I found a couple problem cases to fix by using generated tests instead of manual cases; I'll fix those.

**Comment by pylang** [comment]
The concept is great.   I'm intrigued by how you handled negative steps. +1

**Comment by bbayles** [comment]
I've added more systematic testing and fixed an over-reliance on caching when `step` is negative and `stop` is `None`.

I'll merge unless there are objections - many thanks; this was fun to puzzle through.

I've found a bug in slicing for `peekable` in doing this; I'll make a separate PR for that.

File: more_itertools/more.py [hunk 1]
```diff
@@ -1394,3 +1395,115 @@ def strip(iterable, pred):
 
     """
     return rstrip(lstrip(iterable, pred), pred)
+
+
+def islice_extended(iterable, *args):
+    """An extension of :func:`itertools.islice` that supports negative values
+    for *stop*, *start*, and *step*.
+
+        >>> from more_itertools import ncycles
```
Comment (pylang) [inline more_itertools/more.py]: `ncycles` is imported and not used
